### ISSUE TITLE
error in self attention visualisation 

### ISSUE THREAD

**Issue by twangnh** [facebookresearch/detr#162]
I beleive there is a error for idexing the self attetion:
```
for idx_o, ax in zip(idxs, axs):
    idx = (idx_o[0] // fact, idx_o[1] // fact)
    ax.imshow(sattn[..., idx[0], idx[1]], cmap='cividis', interpolation='nearest')
    ax.axis('off')
    ax.set_title(f'self-attention{idx_o}')
```

should be 
`sattn[idx[0], idx[1], ...]`as the last two dim sums to 1, so the first two are locations and are the dim to indexing, with
this the attention map is 
![image](https://user-images.githubusercontent.com/18298163/88396051-92649100-cdf4-11ea-8583-4e88a84e3597.png)

**Comment by eagleanurag** [comment]
Yes ! we have some other little issues too. Please Pull request regarding your concerns and suggestions

**Comment by twangnh** [comment]
Could you pls share about what other little issues are? thank you.

**Comment by fmassa** [comment]
Hi,

This is a fair point and is discussed in the bottom of the notebook as well
![image](https://user-images.githubusercontent.com/9110200/88709469-7a9f4c80-d115-11ea-9e6f-72ef94b3420c.png)

In the end, it's a matter of deciding if you want the attention of all pixels at location x, or the contribution of location x to all the attention maps.

I would say both information are useful to be considered, and as such I added both in the widget visualization.

**Comment by songhwanjun** [comment]
I also found an error in computing the **scale** **factor**.  When computing the factor, the code currently divides the original image "**height**" with the input image "**width

**".
```
scale = im.height / img.shape[-1]
```
should be changed into
```
scale = im.height / img.shape[-2]
```

Hence, the correct visualization is ![correct_one](https://user-images.githubusercontent.com/10972556/90869514-46950f80-e3d3-11ea-90a1-060728bd28d2.png)

**Comment by TimGor1997** [comment]
> I beleive there is a error for idexing the self attetion:
> 
> ```
> for idx_o, ax in zip(idxs, axs):
>     idx = (idx_o[0] // fact, idx_o[1] // fact)
>     ax.imshow(sattn[..., idx[0], idx[1]], cmap='cividis', interpolation='nearest')
>     ax.axis('off')
>     ax.set_title(f'self-attention{idx_o}')
> ```
> 
> should be
> `sattn[idx[0], idx[1], ...]`as the last two dim sums to 1, so the first two are locations and are the dim to indexing, with
> this the attention map is
> ![image](https://user-images.githubusercontent.com/18298163/88396051-92649100-cdf4-11ea-8583-4e88a84e3597.png)

could you tell me how to "visualize encoder-decoder multi-head attention weights"?
I've copy the code in the pycharm, but nothing show at all……

**Comment by TimGor1997** [comment]
> Yes ! we have some other little issues too. Please Pull request regarding your concerns and suggestions

could you tell me how to "visualize encoder-decoder multi-head attention weights"?
I've copy the code in the pycharm, but nothing show at all……

**Comment by TimGor1997** [comment]
> I also found an error in computing the **scale** **factor**. When computing the factor, the code currently divides the original image "**height**" with the input image "**width**".
> 
> ```
> scale = im.height / img.shape[-1]
> ```
> 
> should be changed into
> 
> ```
> scale = im.height / img.shape[-2]
> ```
> 
> Hence, the correct visualization is
> ![correct_one](https://user-images.githubusercontent.com/10972556/90869728-95db4000-e3d3-11ea-8ef7-8303ef94e7ef.png)

could you tell me how to "visualize encoder-decoder multi-head attention weights"?
I've copy the code in the pycharm, but nothing show at all……

**Comment by jing-yu-lim** [comment]
hi @fmassa  could you kindly explain further what do you mean by '...deciding if you want the attention of all pixels at location x, or the contribution of location x to all the attention maps.' as well as the intuition behind sattn[..., idx[0],idx[1]] and vice versa resulting in instance segmentation? 

Personally the square matrix of self attn weights (HW x HW) is more intuitive to me; i understand it as the result of Query (HW x dim)  and Key.T (dim x HW) matrix mult. I think of row 1 as belonging to Query 1, and each value in the vector corresponding to a scalar weight Query 1 'thinks' should be applied to every Vector. Thus it blew my mind that the weight matrix can be reshaped, and a reference point can be an input to show what that point is attending to.

Thank you!
